### PR TITLE
New ScoringModel for Google Cloud Vertex AI Ranking API and new Google Cloud Storage document loader

### LINKED ISSUE #1819: [FEATURE] Vertex AI Ranking API added as a LangChain4j ScoringModel
**Is your feature request related to a problem? Please describe.**

Google Cloud Vertex AI offers a [Ranking](https://cloud.google.com/generative-ai-app-builder/docs/ranking) API.
The idea is to implement LangChain4j's `ScoringModel` interface to give access to this API.

### PR DESCRIPTION
**Pull request by glaforge**
## Issue

Closes #1819

## Change

Add support for the Vertex AI Ranking API, by implement a `ScoringModel` for it.
Documentation: https://docs.langchain4j.dev/integrations/scoring-reranking-models/vertex-ai and https://docs.langchain4j.dev/integrations/document-loaders/google-cloud-storage

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [X] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

### CODE REVIEW

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+
+/**
+ * Implementation of a <code>ScoringModel</code> for the Vertex AI Ranking API:
+ * https://cloud.google.com/generative-ai-app-builder/docs/ranking
+ */
+public class VertexAiRankerScoringModel implements ScoringModel {
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]:
```suggestion
public class VertexAiScoringModel implements ScoringModel {
```
Do you ming renaming it?

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            "Where sunlight learns to dance and play.\n" +
+            "Blue, a hue of scattered light,\n" +
+            "A gentle whisper, soft and bright."
+        ).map(TextSegment::from).collect(Collectors.toList()),
+    "Why is the sky blue?");
+
+// [0.8199999928474426, 0.4300000071525574]
+```
+
+By default, the model `semantic-ranker-512@latest` will be used.
```
Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]: We are moving away from default values for model names (as they get outdated very soon lol), can you please make it mandatory?

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+You can score a single string or `TextSegment` against a query 
+thanks to the `score(text, query)` and `score(segment, query)` methods.
+
+It is also possible to score several strings or `TextSegment`s against the query, 
+with the `scoreAll(segments, query)` method:
+
+```java
+VertexAiRankerScoringModel scoringModel = VertexAiRankerScoringModel.builder()
+    .projectId(System.getenv("GCP_PROJECT_ID"))
+    .projectNum(System.getenv("GCP_PROJECT_NUM"))
```
Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]: What is project num?

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                        String.format("projects/%s/locations/%s/rankingConfigs/default_ranking_config.", projectNum, location))
+                    .build().getRankingConfig())
+                .setQuery(query)
+                .setIgnoreRecordDetailsInResponse(true)
+                .addAllRecords(segments.stream()
+                    .map(segment -> {
+                        RankingRecord.Builder rankingBuilder = RankingRecord.newBuilder()
+                            .setContent(segment.text());
+                        // Ranker API takes into account titles in its score calculations
+                        String titleKeyToUse = titleMetadataKey != null ? titleMetadataKey : "title";
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: nit: default value can be set in the constructor

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                        rankingBuilder.setId(String.valueOf(counter.getAndIncrement()));
+                        return rankingBuilder.build();
+                    })
+                    .collect(Collectors.toList()));
+
+            RankResponse rankResponse = rankServiceClient.rank(rankingRequestBuilder.build());
+
+            return Response.from(rankResponse.getRecordsList().stream()
+                .peek(rankingRecord -> {
+                    System.out.println(rankingRecord.toString());
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: please remove this

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                })
+                // the API returns results sorted by relevance score, so reorder them back to original order
+                .sorted((rr1, rr2) -> (Double.valueOf(rr1.getId()) > Double.valueOf(rr2.getId())) ? 1 : -1)
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]:
```suggestion
                .sorted(comparing(rr -> Double.valueOf(rr.getId())))
```

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                .map(RankingRecord::getScore)
+                .map(Double::valueOf)
+                .collect(Collectors.toList()));
+        } catch (IOException e) {
+            throw new RuntimeException("An error happened when instantiating the RankServiceSettings class.", e);
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]:
```suggestion
            throw new RuntimeException(e);
```
nit: Are we sure this is the reason? `IOException` sounds more like a network problem

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            this.model = ensureNotBlank(model, "model");
+            return this;
+        }
+
+        public Builder projectId(String projectId) {
+            this.projectId = projectId;
+            return this;
+        }
+
+        public Builder projectNum(String projectNum) {
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: nit:
```suggestion
        public Builder projectNumber(String projectNumber) {
```

File: langchain4j-vertex-ai/src/test/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModelIT.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                    "wavelengths than other colors, and is thus scattered more easily.",
+
+                "A canvas stretched across the day,\n" +
+                    "Where sunlight learns to dance and play.\n" +
+                    "Blue, a hue of scattered light,\n" +
+                    "A gentle whisper, soft and bright."
+            ).map(TextSegment::from).collect(Collectors.toList()),
+            "Why is the sky blue?");
+
+        System.out.println("score = " + score);
```
Comment (langchain4j) [inline langchain4j-vertex-ai/src/test/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModelIT.java]: nit: please remove sysouts from tests

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java (hunk 1 above)
Comment (glaforge) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: Done. Pushed.

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+### Usage
+
+You can score a single string or `TextSegment` against a query 
+thanks to the `score(text, query)` and `score(segment, query)` methods.
+
+It is also possible to score several strings or `TextSegment`s against the query, 
+with the `scoreAll(segments, query)` method:
+
+```java
+VertexAiRankerScoringModel scoringModel = VertexAiRankerScoringModel.builder()
```
Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]:
```suggestion
VertexAiScoringModel scoringModel = VertexAiScoringModel.builder()
```

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+But you can specify the version of the model you want to use, via the `model()` builder method.
+
+If you pass `TextSegment`s which have a particular `title` key, the Ranker model can take this metadata into account in its calculation.
+To specify a custom title key, you can use the `titleMetadataKey()` builder method.`
+
+You can use scoring models with `AiServices` and its `contentAgregator()` method, 
+which takes a `ContentAggregator` class that can specify a scoring model:
+
+```java
+VertexAiRankerScoringModel scoringModel = VertexAiRankerScoringModel.builder()
```
Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]:
```suggestion
VertexAiScoringModel scoringModel = VertexAiScoringModel.builder()
```

**Review by ghost — COMMENTED** [review]
@glaforge great stuff, thank you!

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md (hunk 2 above)
Comment (glaforge) [reply docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]: It's the behavior of the underlying API that I followed.
But ok, I can make it mandatory.

Comment (langchain4j) [reply docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]: Oh, if model name is optional in the API, then let's keep it optional as well (without default value)

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md (hunk 3 above)
Comment (glaforge) [reply docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]: This API requires also the project _number_ (in addition to the project _id_).

You'll have to add it to your CI environment variables.

The project number can be found in the welcome dashboard in the cloud web console. or you can also run `$ gcloud projects describe langchain4j`, and you'll find it in the output.

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java (hunk 5 above)
Comment (glaforge) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: oops!

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java (hunk 6 above)
Comment (glaforge) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: But using `comparing` that way, it's going to compare strings, as `getId()` returns strings.

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java (hunk 7 above)
Comment (glaforge) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: Yes, a strange exception, but that's what is (potentially) thrown by `RankServiceSettings build()`.

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java (hunk 6 above)
Comment (langchain4j) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: My bad, corrected

File: langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java (hunk 7 above)
Comment (langchain4j) [reply langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiRankerScoringModel.java]: I would still remove the message, it does not really add any value

File: docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md (hunk 3 above)
Comment (langchain4j) [reply docs/docs/integrations/scoring-reranking-models/3-vertex-ai-ranking.md]: It would probably make sense to add this info to the docu?

**Review by ghost — APPROVED** [review]
@glaforge thank you, good job!

**Pull request merged**